// @rg-software/angular-archwizard/esm2022/lib/directives/completed-step.directive.mjs
import { Directive, Host, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../util/wizard-step.interface";
/**
 * The `awCompletedStep` directive can be used to make a wizard step initially completed.
 *
 * Initially completed steps are shown as completed when the wizard is presented to the user.
 *
 * A typical use case is to make a step initially completed if it is automatically filled with some derived/predefined information.
 *
 * ### Syntax
 *
 * ```html
 * <aw-wizard-step awCompletedStep>
 *     ...
 * </aw-wizard-step>
 * ```
 *
 * An optional boolean condition can be specified:
 *
 * ```html
 * <aw-wizard-step [awCompletedStep]="shouldBeCompleted">
 *     ...
 * </aw-wizard-step>
 * ```
 *
 * ### Example
 *
 * ```html
 * <aw-wizard-step stepTitle="First step" [awCompletedStep]="firstStepPrefilled">
 *     ...
 * </aw-wizard-step>
 * ```
 */
export class CompletedStepDirective {
    /**
     * Constructor
     *
     * @param wizardStep The wizard step, which contains this [[CompletedStepDirective]]
     */
    constructor(wizardStep) {
        this.wizardStep = wizardStep;
        // tslint:disable-next-line:no-input-rename
        this.initiallyCompleted = true;
    }
    /**
     * Initialization work
     */
    ngOnInit() {
        // The input receives '' when specified in the template without a value.  In this case, apply the default value (`true`).
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        this.wizardStep.initiallyCompleted = this.initiallyCompleted || this.initiallyCompleted === '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: CompletedStepDirective, deps: [{ token: i1.WizardStep, host: true }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.1.0", type: CompletedStepDirective, selector: "[awCompletedStep]", inputs: { initiallyCompleted: ["awCompletedStep", "initiallyCompleted"] }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: CompletedStepDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[awCompletedStep]'
                }]
        }], ctorParameters: () => [{ type: i1.WizardStep, decorators: [{
                    type: Host
                }] }], propDecorators: { initiallyCompleted: [{
                type: Input,
                args: ['awCompletedStep']
            }] } });
