// @rg-software/angular-archwizard/esm2022/lib/directives/enable-back-links.directive.mjs
import { Directive, EventEmitter, Host, Output } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../util/wizard-completion-step.interface";
/**
 * The `awEnableBackLinks` directive can be used to allow the user to leave a [[WizardCompletionStep]] after is has been entered.
 *
 * ### Syntax
 *
 * ```html
 * <aw-wizard-completion-step awEnableBackLinks (stepExit)="exit function">
 *     ...
 * </aw-wizard-completion-step>
 * ```
 *
 * ### Example
 *
 * ```html
 * <aw-wizard-completion-step stepTitle="Final step" awEnableBackLinks>
 *     ...
 * </aw-wizard-completion-step>
 * ```
 *
 * @author Marc Arndt
 */
export class EnableBackLinksDirective {
    /**
     * Constructor
     *
     * @param completionStep The wizard completion step, which should be exitable
     */
    constructor(completionStep) {
        this.completionStep = completionStep;
        /**
         * This EventEmitter is called when the step is exited.
         * The bound method can be used to do cleanup work.
         */
        this.stepExit = new EventEmitter();
    }
    /**
     * Initialization work
     */
    ngOnInit() {
        this.completionStep.canExit = true;
        this.completionStep.stepExit = this.stepExit;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: EnableBackLinksDirective, deps: [{ token: i1.WizardCompletionStep, host: true }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.1.0", type: EnableBackLinksDirective, selector: "[awEnableBackLinks]", outputs: { stepExit: "stepExit" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: EnableBackLinksDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[awEnableBackLinks]'
                }]
        }], ctorParameters: () => [{ type: i1.WizardCompletionStep, decorators: [{
                    type: Host
                }] }], propDecorators: { stepExit: [{
                type: Output
            }] } });
