// @rg-software/angular-archwizard/esm2022/lib/directives/optional-step.directive.mjs
import { Directive, Host, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../util/wizard-step.interface";
/**
 * The `awOptionalStep` directive can be used to define an optional `wizard-step`.
 * An optional wizard step is a [[WizardStep]] that doesn't need to be completed to transition to later wizard steps.
 *
 * ### Syntax
 *
 * ```html
 * <aw-wizard-step awOptionalStep>
 *     ...
 * </aw-wizard-step>
 * ```
 *
 * ### Example
 *
 * ```html
 * <aw-wizard-step stepTitle="Second step" awOptionalStep>
 *     ...
 * </aw-wizard-step>
 * ```
 *
 * @author Marc Arndt
 */
export class OptionalStepDirective {
    /**
     * Constructor
     *
     * @param wizardStep The wizard step, which contains this [[OptionalStepDirective]]
     */
    constructor(wizardStep) {
        this.wizardStep = wizardStep;
        // tslint:disable-next-line:no-input-rename
        this.optional = true;
    }
    /**
     * Initialization work
     */
    ngOnInit() {
        // The input receives '' when specified in the template without a value.  In this case, apply the default value (`true`).
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        this.wizardStep.optional = this.optional || this.optional === '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: OptionalStepDirective, deps: [{ token: i1.WizardStep, host: true }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.1.0", type: OptionalStepDirective, selector: "[awOptionalStep]", inputs: { optional: ["awOptionalStep", "optional"] }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: OptionalStepDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[awOptionalStep]'
                }]
        }], ctorParameters: () => [{ type: i1.WizardStep, decorators: [{
                    type: Host
                }] }], propDecorators: { optional: [{
                type: Input,
                args: ['awOptionalStep']
            }] } });
