// @rg-software/angular-archwizard/esm2022/lib/directives/selected-step.directive.mjs
import { Directive, Host } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../util/wizard-step.interface";
/**
 * The `awSelectedStep` directive can be used on a [[WizardStep]] to set it as selected after the wizard initialisation or a reset.
 *
 * ### Syntax
 *
 * ```html
 * <aw-wizard-step stepTitle="Step title" awSelectedStep>
 *     ...
 * </aw-wizard-step>
 * ```
 *
 * @author Marc Arndt
 */
export class SelectedStepDirective {
    /**
     * Constructor
     *
     * @param wizardStep The wizard step, which should be selected by default
     */
    constructor(wizardStep) {
        this.wizardStep = wizardStep;
    }
    /**
     * Initialization work
     */
    ngOnInit() {
        this.wizardStep.defaultSelected = true;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: SelectedStepDirective, deps: [{ token: i1.WizardStep, host: true }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.1.0", type: SelectedStepDirective, selector: "[awSelectedStep]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: SelectedStepDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[awSelectedStep]'
                }]
        }], ctorParameters: () => [{ type: i1.WizardStep, decorators: [{
                    type: Host
                }] }] });
