// @rg-software/angular-archwizard/esm2022/lib/archwizard.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { WizardCompletionStepComponent } from './components/wizard-completion-step.component';
import { WizardNavigationBarComponent } from './components/wizard-navigation-bar.component';
import { WizardStepComponent } from './components/wizard-step.component';
import { WizardComponent } from './components/wizard.component';
import { EnableBackLinksDirective } from './directives/enable-back-links.directive';
import { GoToStepDirective } from './directives/go-to-step.directive';
import { NextStepDirective } from './directives/next-step.directive';
import { OptionalStepDirective } from './directives/optional-step.directive';
import { PreviousStepDirective } from './directives/previous-step.directive';
import { ResetWizardDirective } from './directives/reset-wizard.directive';
import { SelectedStepDirective } from './directives/selected-step.directive';
import { WizardCompletionStepDirective } from './directives/wizard-completion-step.directive';
import { WizardStepSymbolDirective } from './directives/wizard-step-symbol.directive';
import { WizardStepTitleDirective } from './directives/wizard-step-title.directive';
import { WizardStepDirective } from './directives/wizard-step.directive';
import { NavigationModeDirective } from './directives/navigation-mode.directive';
import { CompletedStepDirective } from './directives/completed-step.directive';
import * as i0 from "@angular/core";
/**
 * The module defining all the content inside `angular-archwizard`
 *
 * @author Marc Arndt
 */
export class ArchwizardModule {
    /* istanbul ignore next */
    static forRoot() {
        return {
            ngModule: ArchwizardModule,
            providers: [
            // Nothing here yet
            ]
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: ArchwizardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.1.0", ngImport: i0, type: ArchwizardModule, declarations: [WizardComponent,
            WizardStepComponent,
            WizardNavigationBarComponent,
            WizardCompletionStepComponent,
            GoToStepDirective,
            NextStepDirective,
            PreviousStepDirective,
            OptionalStepDirective,
            WizardStepSymbolDirective,
            WizardStepTitleDirective,
            EnableBackLinksDirective,
            WizardStepDirective,
            WizardCompletionStepDirective,
            SelectedStepDirective,
            ResetWizardDirective,
            NavigationModeDirective,
            CompletedStepDirective], imports: [CommonModule], exports: [WizardComponent,
            WizardStepComponent,
            WizardNavigationBarComponent,
            WizardCompletionStepComponent,
            GoToStepDirective,
            NextStepDirective,
            PreviousStepDirective,
            OptionalStepDirective,
            WizardStepSymbolDirective,
            WizardStepTitleDirective,
            EnableBackLinksDirective,
            WizardStepDirective,
            WizardCompletionStepDirective,
            SelectedStepDirective,
            ResetWizardDirective,
            NavigationModeDirective,
            CompletedStepDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: ArchwizardModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.0", ngImport: i0, type: ArchwizardModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        WizardComponent,
                        WizardStepComponent,
                        WizardNavigationBarComponent,
                        WizardCompletionStepComponent,
                        GoToStepDirective,
                        NextStepDirective,
                        PreviousStepDirective,
                        OptionalStepDirective,
                        WizardStepSymbolDirective,
                        WizardStepTitleDirective,
                        EnableBackLinksDirective,
                        WizardStepDirective,
                        WizardCompletionStepDirective,
                        SelectedStepDirective,
                        ResetWizardDirective,
                        NavigationModeDirective,
                        CompletedStepDirective,
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        WizardComponent,
                        WizardStepComponent,
                        WizardNavigationBarComponent,
                        WizardCompletionStepComponent,
                        GoToStepDirective,
                        NextStepDirective,
                        PreviousStepDirective,
                        OptionalStepDirective,
                        WizardStepSymbolDirective,
                        WizardStepTitleDirective,
                        EnableBackLinksDirective,
                        WizardStepDirective,
                        WizardCompletionStepDirective,
                        SelectedStepDirective,
                        ResetWizardDirective,
                        NavigationModeDirective,
                        CompletedStepDirective,
                    ]
                }]
        }] });
