// @siemens/ngx-datatable/esm2022/lib/components/body/body-row-wrapper.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, HostListener, inject, Input, Output, ViewChild } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { DatatableComponentToken } from '../../utils/table-token';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class DataTableRowWrapperComponent {
    set rowIndex(val) {
        this._rowIndex = val;
        this.rowContext.rowIndex = val;
        this.groupContext.rowIndex = val;
        this.cd.markForCheck();
    }
    get rowIndex() {
        return this._rowIndex;
    }
    set expanded(val) {
        this._expanded = val;
        this.groupContext.expanded = val;
        this.rowContext.expanded = val;
        this.cd.markForCheck();
    }
    get expanded() {
        return this._expanded;
    }
    constructor(cd, differs, iterableDiffers) {
        this.cd = cd;
        this.differs = differs;
        this.iterableDiffers = iterableDiffers;
        this.rowContextmenu = new EventEmitter(false);
        this.selectedGroupRows = [];
        this._expanded = false;
        this.tableComponent = inject(DatatableComponentToken);
        this.groupContext = {
            group: this.row,
            expanded: this.expanded,
            rowIndex: this.rowIndex
        };
        this.rowContext = {
            row: this.row,
            expanded: this.expanded,
            rowIndex: this.rowIndex
        };
        this.rowDiffer = differs.find({}).create();
        this.selectedRowsDiffer = this.iterableDiffers.find(this.selected ?? []).create();
    }
    ngOnInit() {
        if (this.disableCheck) {
            const isRowDisabled = this.disableCheck(this.row);
            this.disable$ = new BehaviorSubject(isRowDisabled);
        }
        this.rowContext.disableRow$ = this.disable$;
    }
    ngDoCheck() {
        if (this.disableCheck) {
            const isRowDisabled = this.disableCheck(this.row);
            this.disable$.next(isRowDisabled);
            this.cd.markForCheck();
        }
        if (this.rowDiffer.diff(this.row)) {
            this.rowContext.row = this.row;
            this.groupContext.group = this.row;
            this.cd.markForCheck();
        }
        // When groupheader is used with chechbox we use iterableDiffer
        // on currently selected rows to check if it is modified
        // if any of the row of this group is not present in `selected` rows array
        // mark group header checkbox state as indeterminate
        if (this.groupHeader?.checkboxable && this.selectedRowsDiffer.diff(this.selected)) {
            const selectedRows = this.selected.filter(row => this.row.value.find(item => item === row));
            if (this.checkBoxInput) {
                if (selectedRows.length && selectedRows.length !== this.row.value.length) {
                    this.checkBoxInput.nativeElement.indeterminate = true;
                }
                else {
                    this.checkBoxInput.nativeElement.indeterminate = false;
                }
            }
            this.selectedGroupRows = selectedRows;
        }
    }
    onContextmenu($event) {
        this.rowContextmenu.emit({ event: $event, row: this.row });
    }
    getGroupHeaderStyle() {
        const styles = {};
        styles.transform = 'translate3d(' + this.offsetX + 'px, 0px, 0px)';
        styles['backface-visibility'] = 'hidden';
        styles.width = this.innerWidth + 'px';
        return styles;
    }
    onCheckboxChange(groupSelected) {
        // First remove all rows of this group from `selected`
        this.selected = [...this.selected.filter(row => !this.row.value.find(item => item === row))];
        // If checkbox is checked then add all rows of this group in `selected`
        if (groupSelected) {
            this.selected = [...this.selected, ...this.row.value];
        }
        // Update `selected` of DatatableComponent with newly evaluated `selected`
        this.tableComponent.selected = [...this.selected];
        // Emit select event with updated values
        this.tableComponent.onBodySelect({
            selected: this.selected
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.6", ngImport: i0, type: DataTableRowWrapperComponent, deps: [{ token: i0.ChangeDetectorRef }, { token: i0.KeyValueDiffers }, { token: i0.IterableDiffers }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.0.6", type: DataTableRowWrapperComponent, selector: "datatable-row-wrapper", inputs: { innerWidth: "innerWidth", rowDetail: "rowDetail", groupHeader: "groupHeader", offsetX: "offsetX", detailRowHeight: "detailRowHeight", groupHeaderRowHeight: "groupHeaderRowHeight", row: "row", groupedRows: "groupedRows", disableCheck: "disableCheck", selected: "selected", rowIndex: "rowIndex", expanded: "expanded" }, outputs: { rowContextmenu: "rowContextmenu" }, host: { listeners: { "contextmenu": "onContextmenu($event)" }, classAttribute: "datatable-row-wrapper" }, viewQueries: [{ propertyName: "checkBoxInput", first: true, predicate: ["select"], descendants: true }], ngImport: i0, template: `
    <div
      *ngIf="groupHeader && groupHeader.template"
      [style.height.px]="groupHeaderRowHeight"
      class="datatable-group-header"
      [ngStyle]="getGroupHeaderStyle()"
    >
      <div class="datatable-group-cell">
        <div *ngIf="groupHeader.checkboxable">
          <label
            class="datatable-checkbox"
          >
            <input #select type="checkbox" [checked]="selectedGroupRows.length === row.value.length" (change)="onCheckboxChange(select.checked)" />
          </label>
        </div>
        <ng-template
          *ngIf="groupHeader && groupHeader.template"
          [ngTemplateOutlet]="groupHeader.template"
          [ngTemplateOutletContext]="groupContext"
        >
        </ng-template>
      </div>
    </div>
    <ng-content *ngIf="(groupHeader && groupHeader.template && expanded) || !groupHeader || !groupHeader.template">
    </ng-content>
    <div
      *ngIf="rowDetail && rowDetail.template && expanded"
      [style.height.px]="detailRowHeight"
      class="datatable-row-detail"
    >
      <ng-template
        *ngIf="rowDetail && rowDetail.template"
        [ngTemplateOutlet]="rowDetail.template"
        [ngTemplateOutletContext]="rowContext"
      >
      </ng-template>
    </div>
  `, isInline: true, dependencies: [{ kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i1.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }, { kind: "directive", type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.6", ngImport: i0, type: DataTableRowWrapperComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'datatable-row-wrapper',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: `
    <div
      *ngIf="groupHeader && groupHeader.template"
      [style.height.px]="groupHeaderRowHeight"
      class="datatable-group-header"
      [ngStyle]="getGroupHeaderStyle()"
    >
      <div class="datatable-group-cell">
        <div *ngIf="groupHeader.checkboxable">
          <label
            class="datatable-checkbox"
          >
            <input #select type="checkbox" [checked]="selectedGroupRows.length === row.value.length" (change)="onCheckboxChange(select.checked)" />
          </label>
        </div>
        <ng-template
          *ngIf="groupHeader && groupHeader.template"
          [ngTemplateOutlet]="groupHeader.template"
          [ngTemplateOutletContext]="groupContext"
        >
        </ng-template>
      </div>
    </div>
    <ng-content *ngIf="(groupHeader && groupHeader.template && expanded) || !groupHeader || !groupHeader.template">
    </ng-content>
    <div
      *ngIf="rowDetail && rowDetail.template && expanded"
      [style.height.px]="detailRowHeight"
      class="datatable-row-detail"
    >
      <ng-template
        *ngIf="rowDetail && rowDetail.template"
        [ngTemplateOutlet]="rowDetail.template"
        [ngTemplateOutletContext]="rowContext"
      >
      </ng-template>
    </div>
  `,
                    host: {
                        class: 'datatable-row-wrapper'
                    }
                }]
        }], ctorParameters: () => [{ type: i0.ChangeDetectorRef }, { type: i0.KeyValueDiffers }, { type: i0.IterableDiffers }], propDecorators: { checkBoxInput: [{
                type: ViewChild,
                args: ['select']
            }], innerWidth: [{
                type: Input
            }], rowDetail: [{
                type: Input
            }], groupHeader: [{
                type: Input
            }], offsetX: [{
                type: Input
            }], detailRowHeight: [{
                type: Input
            }], groupHeaderRowHeight: [{
                type: Input
            }], row: [{
                type: Input
            }], groupedRows: [{
                type: Input
            }], disableCheck: [{
                type: Input
            }], selected: [{
                type: Input
            }], rowContextmenu: [{
                type: Output
            }], rowIndex: [{
                type: Input
            }], expanded: [{
                type: Input
            }], onContextmenu: [{
                type: HostListener,
                args: ['contextmenu', ['$event']]
            }] } });
